diff --git a/build.zig b/build.zig
index 6d1fdb5..9d2e4fd 100644
--- a/build.zig
+++ b/build.zig
@@ -25,7 +25,7 @@ pub fn build(b: *std.Build) void {
                 &[_][]const u8{ "git", "-C", b.build_root.path orelse ".", "describe", "--long" },
                 &ret,
-                .Inherit,
+                .inherit,
             ) catch break :blk version;
@@ -79,7 +79,7 @@ pub fn build(b: *std.Build) void {
 
     zelbar.root_module.addOptions("build_options", options);
 
-    zelbar.linkLibC();
+    zelbar.link_libc = true;
 
     zelbar.root_module.addImport("wayland", wayland);
